packages/medusa/src/strategies/cart-completion.ts

```diff
@@ -326,6 +156,182 @@ class CartCompletionStrategy extends AbstractCartCompletionStrategy {
+        "shipping_address",
+        "shipping_methods",
+      ],
+    })
+
+    if (cart.completed_at) {
+      return {
+        response_code: 409,
+        response_body: {
+          code: MedusaError.Codes.CART_INCOMPATIBLE_STATE,
+          message: "Cart has already been completed",
+          type: MedusaError.Types.NOT_ALLOWED,
+        },
+      }
+    }
+
+    await this.cartService_.withTransaction(manager).createTaxLines(cart)
+
+    return {
+      recovery_point: "tax_lines_created",
+    }
+  }
+
+  protected async handleTaxLineCreated(
+    id: string,
+    idempotencyKey: IdempotencyKey,
+    { context, manager }: { context: any; manager: EntityManager }
+  ) {
+    const cart = await this.cartService_
+      .withTransaction(manager)
+      .authorizePayment(id, {
+        ...context,
+        cart_id: id,
+        idempotency_key: idempotencyKey,
+        useExistingTaxLine: true,
+      })
+
+    if (cart.payment_session) {
+      if (
+        cart.payment_session.status === "requires_more" ||
+        cart.payment_session.status === "pending"
+      ) {
+        await this.cartService_.withTransaction(manager).deleteTaxLines(id)
+
+        return {
+          response_code: 200,
+          response_body: {
+            data: cart,
+            payment_status: cart.payment_session.status,
+            type: "cart",
+          },
+        }
+      }
+    }
+
+    return {
+      recovery_point: "payment_authorized",
+    }
+  }
+
+  protected async handlePaymentAuthorized(
+    id: string,
+    { manager }: { manager: EntityManager }
+  ) {
+    const orderServiceTx = this.orderService_.withTransaction(manager)
+
+    const cart = await this.cartService_
+      .withTransaction(manager)
+      .retrieveWithTotals(id, {
+        relations: ["region", "payment", "payment_sessions"],
+      })
+
+    // If cart is part of swap, we register swap as complete
+    if (cart.type === "swap") {
+      try {
+        const swapId = cart.metadata?.swap_id
+        let swap = await this.swapService_
+          .withTransaction(manager)
+          .registerCartCompletion(swapId as string)
+
+        swap = await this.swapService_
+          .withTransaction(manager)
+          .retrieve(swap.id, {
+            relations: ["shipping_address"],
+          })
+
+        return {
+          response_code: 200,
+          response_body: { data: swap, type: "swap" },
+        }
+      } catch (error) {
+        if (error && error.code === MedusaError.Codes.INSUFFICIENT_INVENTORY) {
+          return {
+            response_code: 409,
+            response_body: {
+              message: error.message,
+              type: error.type,
+              code: error.code,
+            },
+          }
+        } else {
+          throw error
+        }
+      }
+    }
+
+    if (!isDefined(cart.total)) {
+      return {
+        response_code: 500,
+        response_body: {
+          message: "Unexpected state",
+        },
+      }
+    }
```

Comment:
thought: I think this seems a bit arbitrary. Can we add a comment and improve the error message to provide a better experience for both developers reading the code as well as the ones who are consuming the API?

Reply:
I should be able to remove it as it is a legacy error which should not happen anymore with the new decorate totals method